include/ldsCtrlEst_h/lds_mpcctrl.h

```diff
@@ -0,0 +1,533 @@
+  //   std::cout << "[" << matrix->m << ", " << matrix->n << "]:\n";
+
+  //   for (int j = 0, k = 0; j < (matrix->p[matrix->n]); j++) {
+  //     while (k <= matrix->n && matrix->p[k + 1] <= j) k++;
+  //     std::cout << "(" << matrix->i[j]  << ", " << k << ") " << matrix->x[j] << "\n";
+  //   }
+  //   std::cout << std::endl;
+  // }
+
+  void Init() {
+    A_ = sys_.A();
+    B_ = sys_.B();
+    n_ = B_.n_rows;
+    m_ = B_.n_cols;
+    xi_ = arma::zeros<Vector>(n_);
+    t_sim_ = 0;
+  }
+};
+
+// Implement methods
+
+template <typename System>
+inline MpcController<System>::MpcController(const System& sys, Vector u_lb,
+                                            Vector u_ub)
+    : sys_(sys), lb_(u_lb), ub_(u_ub) {
+  Init();
+}
+
+template <typename System>
+inline MpcController<System>::MpcController(System&& sys, Vector u_lb,
+                                            Vector u_ub)
+    : sys_(std::move(sys)), lb_(u_lb), ub_(u_ub) {
+  Init();
+}
+
+template <typename System>
+Vector MpcController<System>::Control(data_t t_sim, const Vector& x0,
+                                      const Vector& u0, const Matrix& xr) {
+  size_t n_sim = t_sim / sys_.dt();  // Number of points per simulation step
+
+  OSQPSolution* sol;
+  if (arma::all(xi_ == x0) && (t_sim_ == t_sim)) {
+    sol = fast_update(x0, u0, xr, n_sim);
+  } else {
+    sol = slow_update(x0, u0, xr, n_sim);
+  }
+  t_sim_ = t_sim;
+
+  Vector ui(m_);
+  for (int i = 0; i < m_; i++) {
+    ui(i) = sol->x[N_ * n_ + i];
+  }
+
+  if (sol) free(sol);
+
+  return ui;
+}
+
+template <typename System>
+OSQPSolution* MpcController<System>::fast_update(const Vector& x0,
+                                                 const Vector& u0,
+                                                 const Matrix& xr,
+                                                 size_t n_sim) {
+  lb_.rows(0, n_ - 1) = -x0.t();
+  ub_.rows(0, n_ - 1) = -x0.t();
+
+  // Convert state penalty from reference to OSQP format
+  arma::Row<data_t> q_arma;
+  {
+    arma::uvec indices = arma::regspace<arma::uvec>(0, n_sim, N_ * n_sim - 1);
+    Matrix sliced_xr = xr.cols(indices);
+    Matrix Qxr_full = -2 * Q_ * sliced_xr;
+    arma::Row<data_t> Qxr =
+        Qxr_full.as_col().t();  // Qxr for every simulation time step
+
+    arma::Row<data_t> qu =
+        join_horiz(-2 * S_ * u0, arma::zeros<Vector>((M_ - 1) * m_));
+    arma::Row<data_t> qx = Qxr.cols(0, N_ * n_ - 1);
+    q_arma = join_horiz(qx, qu);
+  }
+
+  // Solver, settings, matrices
+  OSQPSolver* solver;
+  OSQPSettings* settings;
+  OSQPCscMatrix* A = from_matrix(Acon_);
+  OSQPCscMatrix* P = from_matrix(P_);
+  OSQPFloat* q = from_vec(q_arma.t());
+  OSQPFloat* lb = from_vec(lb_);
+  OSQPFloat* ub = from_vec(ub_);
+
+  // Set settings
+  settings = (OSQPSettings*)malloc(sizeof(OSQPSettings));
+  if (settings) {
+    osqp_set_default_settings(settings);
+    settings->verbose = false;
+  }
```

Comment:
I see that this follow the example of Python lqmpc, where prob.setup() is called every step. But is there a way to keep the solver, settings, matrices, vectors, etc. around from one step to the next? I imagine this would help things go faster, and might help us keep memory leaks under control since pointers wouldn't be created on every step

Reply:
I didn't see this comment before I responded to the previous one, but I think that is probably what I should be aiming for and I will try my best to reuse as many of the variables as possible.